### PR TITLE
Added Window Position for LJWGL3 libraries

### PR DESCRIPTION
**Pull request by bob0bob**
This is a very simple addition. It allows a person to set 3 variables in AppSettings. ‘CenterWindow’, ‘WindowXPosition’ and ‘WindowYPosition’ variables. This way these variable will be saved in the profile when the profile is saved, and be reloaded. I added ‘CenterWindow’ to be added with a ‘true’ value for the default value so it will run just like it did before.

But if you set ‘CenterWindow’ to ‘false’ then inside LwjglWindow.java (lwjgl3 code) it will look at these new values, it will determine to center the window or use the position values to place the window back at the location the user last moved it to.

Of course, these values are only updated if the “program” updates this value. So if you want to save screen position, you can save them on closing to and on restart put the window back into the same location.

### CODE REVIEW

**Comment by stephengold** [comment]
Can we close #1668?

**Comment by bob0bob** [comment]
I closed it.

File: jme3-core/src/main/java/com/jme3/system/AppSettings.java [hunk 1]
```diff
@@ -1350,4 +1351,65 @@ public boolean isUseRetinaFrameBuffer() {
...
+     *
+     * @param center whether to center the window or not.
+     */
+    public void setCenterWindow(boolean center) {
+        putBoolean("CenterWindow", center);
+    }
+
+    
+    /**
+     * Get the position of the window's X Position
```
Comment (stephengold) [inline jme3-core/src/main/java/com/jme3/system/AppSettings.java:1376]: The repetition of the word "position" makes this javadoc confusing.
Javadoc should mention that this is only valid for a non-full-screen window with CenterWindow set false.
Javadoc should mention that the X position is that of the upper-left corner of the content area, measured rightward from the left edge of the display.

File: jme3-core/src/main/java/com/jme3/system/AppSettings.java [hunk 2]
```diff
@@ -266,6 +266,7 @@
     public static final String JOAL = "JOAL";
 
     static {
+        defaults.put("CenterWindow", true);
```
Comment (stephengold) [inline jme3-core/src/main/java/com/jme3/system/AppSettings.java:269]: Perhaps we should supply defaults for WindowXPos and WindowYPos, in case the developer sets CenterWindow but not WindowXPos and WindowYPos?

File: jme3-core/src/main/java/com/jme3/system/AppSettings.java [hunk 3]
```diff
@@ -1350,4 +1351,65 @@ public boolean isUseRetinaFrameBuffer() {
...
+     *
+     * @return int screen coordinates of the X Position of the window
+     * @see #setWindowXPosition(int)
+     */
+    public int getWindowXPosition() {
+        return getInteger("windowXpos");
+    }
+    
+    /**
+     * The position of the window's X Position.
```
Comment (stephengold) [inline jme3-core/src/main/java/com/jme3/system/AppSettings.java:1386]: See my review comments on `getWindowXPosition()`.

File: jme3-core/src/main/java/com/jme3/system/AppSettings.java [hunk 4]
```diff
@@ -1350,4 +1351,65 @@ public boolean isUseRetinaFrameBuffer() {
...
+
+    
+    /**
+     * Get the position of the window's X Position
+     *
+     * @return int screen coordinates of the X Position of the window
+     * @see #setWindowXPosition(int)
+     */
+    public int getWindowXPosition() {
+        return getInteger("windowXpos");
```
Comment (stephengold) [inline jme3-core/src/main/java/com/jme3/system/AppSettings.java:1382]: For consistency with the other settings, I think "windowXpos" should be capitalized and spelled out: "WindowXPosition".

File: jme3-core/src/main/java/com/jme3/system/AppSettings.java [hunk 5]
```diff
@@ -1350,4 +1351,65 @@ public boolean isUseRetinaFrameBuffer() {
...
+     *
+     * @param pos screen coordinates for the X position.
+     */
+    public void setWindowXPosition(int pos) {
+        putInteger("windowXpos", pos);
+    }
+
+
+    /**
+     * Get the position of the window's Y Position
```
Comment (stephengold) [inline jme3-core/src/main/java/com/jme3/system/AppSettings.java:1396]: See my review comments on getWindowXPosition().

File: jme3-core/src/main/java/com/jme3/system/AppSettings.java [hunk 6]
```diff
@@ -1350,4 +1351,65 @@ public boolean isUseRetinaFrameBuffer() {
...
+     *
+     * @return int screen coordinates of the Y Position of the window
+     * @see #setWindowXPosition(int)
+     */
+    public int getWindowYPosition() {
+        return getInteger("windowYpos");
+    }
+    
+    /**
+     * The position of the window's Y Position.
```
Comment (stephengold) [inline jme3-core/src/main/java/com/jme3/system/AppSettings.java:1406]: See my review comments on getWindowXPosition().

File: jme3-lwjgl3/src/main/java/com/jme3/system/lwjgl/LwjglWindow.java [hunk 7]
```diff
@@ -276,11 +276,16 @@ public void invoke(final long window, final boolean focus) {
             }
         });
 
-        // Center the window
         if (!settings.isFullscreen()) {
-            glfwSetWindowPos(window,
+            if (settings.getCenterWindow())
+            {
```
Comment (stephengold) [inline jme3-lwjgl3/src/main/java/com/jme3/system/lwjgl/LwjglWindow.java:281]: This curly brace belongs on the previous line of code.

File: jme3-lwjgl3/src/main/java/com/jme3/system/lwjgl/LwjglWindow.java [hunk 8]
```diff
@@ -276,11 +276,16 @@ public void invoke(final long window, final boolean focus) {
...
+                // Center the window
+                glfwSetWindowPos(window,
                     (videoMode.width() - settings.getWidth()) / 2,
                     (videoMode.height() - settings.getHeight()) / 2);
+            } else {
+            	glfwSetWindowPos(window,settings.getWindowXPosition(), settings.getWindowYPosition());
```
Comment (stephengold) [inline jme3-lwjgl3/src/main/java/com/jme3/system/lwjgl/LwjglWindow.java:287]: This line of code should be indented at least 3 spaces further.

**Review by stephengold — CHANGES_REQUESTED** [review]
The code itself looks fine. Most of my concerns center on the javadoc. If you're unsure how to proceed, just ask.

File: jme3-core/src/main/java/com/jme3/system/AppSettings.java (hunk 4 above)
Comment (stephengold) [reply jme3-core/src/main/java/com/jme3/system/AppSettings.java:1386]: What do you think bob0bob?

File: jme3-lwjgl3/src/main/java/com/jme3/system/lwjgl/LwjglWindow.java (hunk 8 above)
Comment (stephengold) [reply jme3-lwjgl3/src/main/java/com/jme3/system/lwjgl/LwjglWindow.java:286]: What do you think bob0bob?

File: jme3-lwjgl3/src/main/java/com/jme3/system/lwjgl/LwjglWindow.java [hunk 9]
```diff
@@ -474,7 +478,8 @@ public void create(boolean waitFor) {
         // NOTE: this is required for Mac OS X!
         mainThread = Thread.currentThread();
         mainThread.setName("jME3 Main");
-        run();
+        new Thread(this, "jME3 Main").start();
```
Comment (stephengold) [inline jme3-lwjgl3/src/main/java/com/jme3/system/lwjgl/LwjglWindow.java:481]: These changes seem unrelated to the issue addressed by this PR.

Comment (bob0bob) [reply jme3-lwjgl3/src/main/java/com/jme3/system/lwjgl/LwjglWindow.java:481]: NOt sure how to stop that.  it automatically pulled them in from my fork.  Do you know how to limit what changed to pull in.

File: jme3-lwjgl3/src/main/java/com/jme3/system/lwjgl/LwjglWindow.java (hunk 8 above)
Comment (bob0bob) [reply jme3-lwjgl3/src/main/java/com/jme3/system/lwjgl/LwjglWindow.java:286]: What are your default tabs.  The tabs are right on my system.

File: jme3-core/src/main/java/com/jme3/system/AppSettings.java (hunk 4 above)
Comment (bob0bob) [reply jme3-core/src/main/java/com/jme3/system/AppSettings.java]: I changed it.

**Comment by bob0bob** [comment]
I got that last change removed that doesn't relate to this issue.

File: jme3-core/src/main/java/com/jme3/system/AppSettings.java (hunk 4 above)
Comment (stephengold) [reply jme3-core/src/main/java/com/jme3/system/AppSettings.java]: It's still starting with a lower-case "w" somehow.

File: jme3-core/src/main/java/com/jme3/system/AppSettings.java (hunk 1 above)
Comment (stephengold) [reply jme3-core/src/main/java/com/jme3/system/AppSettings.java]: Javadoc should mention that the X position is that of the upper-left corner of the content area, measured rightward from the left edge of the display.

**Comment by stephengold** [comment]
Thanks for persevering through the difficulties. There are just a few more review comments to address.

**Comment by stephengold** [comment]
Note that JME does not use tabs in Java sourcecode. Indentation should consist entirely of spaces.

**Review by stephengold — APPROVED** [review]

**Comment by bob0bob** [comment]
I forgot to do this, thanks for updating the comments.

**Comment by stephengold** [comment]
Thanks for your contribution to the project.
